plugins/inputs/elasticsearch_query/aggregation_query.go

```diff
@@ -0,0 +1,203 @@
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	elastic "gopkg.in/olivere/elastic.v5"
+)
+
+type aggKey struct {
+	measurement string
+	name        string
+	function    string
+	field       string
+}
+
+type aggregationQueryData struct {
+	aggKey
+	isParent    bool
+	aggregation elastic.Aggregation
+}
+
+func (e *ElasticsearchQuery) runAggregationQuery(ctx context.Context, aggregation esAggregation) (*elastic.SearchResult, error) {
+
+	now := time.Now().UTC()
+	from := now.Add(aggregation.QueryPeriod.Duration * -1)
+	filterQuery := aggregation.FilterQuery
+
+	if filterQuery == "" {
+		filterQuery = "*"
+	}
+
+	query := elastic.NewBoolQuery()
+	query = query.Filter(elastic.NewQueryStringQuery(filterQuery))
+	query = query.Filter(elastic.NewRangeQuery(aggregation.DateField).From(from).To(now))
+
+	search := e.esClient.Search().Index(aggregation.Index).Query(query).Size(0)
+
+	// add only parent elastic.Aggregations to the search request, all the rest are subaggregations of these
+	for _, v := range aggregation.aggregationQueryList {
+		if v.isParent && v.aggregation != nil {
+			search.Aggregation(v.aggKey.name, v.aggregation)
+		}
+	}
+	searchResult, err := search.Do(ctx)
+
+	if err != nil && searchResult != nil {
+		return searchResult, fmt.Errorf("%s - %s", searchResult.Error.Type, searchResult.Error.Reason)
+	}
+
+	return searchResult, err
+
+}
+
+// GetMetricFields function returns a map of fields and field types on Elasticsearch that matches field.MetricFields
+func (e *ElasticsearchQuery) getMetricFields(ctx context.Context, aggregation esAggregation) (map[string]string, error) {
+	mapMetricFields := make(map[string]string)
+
+	if e.esClient == nil {
+		err := e.connectToES()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	for _, metricField := range aggregation.MetricFields {
+
+		resp, err := e.esClient.GetFieldMapping().Index(aggregation.Index).Field(metricField).Do(ctx)
+
+		if err != nil {
+			return mapMetricFields, fmt.Errorf("error retrieving field mappings for %s: %s", aggregation.Index, err.Error())
+		}
+
+		for _, index := range resp {
+			if mappings, ok := index.(map[string]interface{})["mappings"]; ok {
+				if types, ok := mappings.(map[string]interface{}); ok {
+					for _, _type := range types {
+						if field, ok := _type.(map[string]interface{}); ok {
+							for _, _field := range field {
+								if fullname, ok := _field.(map[string]interface{})["full_name"]; ok {
+									if fname, ok := fullname.(string); ok {
+										if mapping, ok := _field.(map[string]interface{})["mapping"]; ok {
+											if fieldType, ok := mapping.(map[string]interface{}); ok {
+												for _, fieldType := range fieldType {
+													if t, ok := fieldType.(map[string]interface{})["type"]; ok {
+														if ftype, ok := t.(string); ok {
+															mapMetricFields[fname] = ftype
+														}
+													}
+												}
+											}
+										}
+									}
+								}
+							}
+						}
+					}
+				}
+			}
```

Comment:
Please try to invert the logic to save indention levels! Furthermore, please check the type assertions, especially for `index` (first `if`).

Reply:
Yes, this does not look good... 
I tried reverting at first, but having to initialize each variable before assigning the value after the assertion check wasn't looking good either, unless I'm missing something.

Reply:
Your new solution is really nice!